[Proposal] - Base class for FeedbackLoops systems.md (#228)
Babylonpartners/ios-playbook: Cookbook/Proposals/Base class for FeedbackLoops systems.md

# Base class for FeedbackLoops systems

* Author(s): Sergey Shulga
* Review Manager: Anders Ha

## Introduction

> Almost for all our screens, we implement ViewModels which are powered by the [ReactiveFeeedback](https://github.com/Babylonpartners/ReactiveFeedback).

Reviewer: ```suggestion
Almost for all our screens, we implement ViewModels which are powered by [ReactiveFeeedback](https://github.com/Babylonpartners/ReactiveFeedback).
```

> The current interface of VM looks like this

Reviewer 2: so how will it change? Will `state.mutate` be the only change with this approach or the way we initialise state will also change? I suppose it will because the renderer signature changes now.

Author: Could be very likely that we won't have `Renderer` any more

Reviewer 2: what we will have instead then?

Author: ```
protocol View: Renderable where Self.View == Never {
    func body() -> AnyRenderable
}

Reviewer 2: I don't think this is related to this proposal

```swift
public protocol BoxViewModel: AnyObject {
    associatedtype State
    associatedtype Action

    var state: Property<State> { get }

    func send(action: Action)
}
```

## Motivation

One of the main issues with this approach

1. We need to implement `input` feedback in pretty much every VM which may lead to errors if do not schedule input correctly, for this we have `Feedback.input()` convenience extension

> 2. If the screen contains a lot of input field we end up having a lot of event's that are just mutate one property of the `State` e.g

Reviewer: ```suggestion
2. If the screen contains a lot of input fields we end up having a lot of events that just mutate one property of the `State` e.g
```

`Action.didChangeUserName(String)` `Action.didPickBirthDate(Date)`
 
## Proposed solution
1. Implementation of the feedback:

```swift

> open class FeedbackLoop<State, Event, Action>: FeedbackLoopType {

Reviewer 2: can you add example of how the subclass of this will look like and how it will be used in a some view model?

Author: ```
init(...) {
super.init(
    initial: State(),
    reducer: VM.reduce,
    scheduler: scheduler,
    feedback: [feedbacks]
)
}
```

Reviewer 2: So the view model will be a subclass of `FeedbackLoop`?

Author: yes

Reviewer 2: I think these kind of questions should be addressed with extensive code snippets in the document so that we can see the whole picture

  public let state: Property<State>
  private let input = Feedback<State, Update>.input()
  private let actionTransformer: (Action) -> Event

  public init(
    initial: State,
    reducer: @escaping (State, Event) -> State,
    feedbacks: [Feedback<State, Event>],
    scheduler: DateScheduler,
    actionTransformer: @escaping (Action) -> Event
  ) {
    self.actionTransformer = actionTransformer
    self.state = Property(
      initial: initial,
      scheduler: scheduler,
      reduce: { (state: State, update: Update) -> State in
        switch update {
        case let .event(event):
          return reducer(state, event)
        case let .mutation(mutation):
          return mutation.mutate(state)
        }

>       },

Reviewer 3: This feels like a higher order reducer.

```
func mutationReducer<State, Event>(_ reducer: (State, Event) -> State) -> (State, Event) -> State {
    return { state, event in 
       switch event {
              case let .event(anEvent):
                     return reducer(state, anEvent)
              case let .mutation(mutation):
                     return mutation.mutate(state)
              }
    }
}
```

And then:

```swift
self.state = Property(
	      initial: initial,
	      scheduler: scheduler,
	      reduce: mutationReducer(reducer),
              feedbacks: feedbacks.map { $0.mapEvent(Update.event) } 
```

Reviewer 4: Yes, it's a lift from `Event` to `Update`:

```swift
func mutationReducer<State, Event>(_ reducer: (State, Event) -> State) 
    -> (State, Update) -> State
```

But what `(State, Update) -> State` basically means is that we will have a **direct way to mutating sub-state at anytime**, which is against the concept of (well-designed) state machine that has the minimal state transition arrows.
If we allow `Update` to be passed in, it means we are going to draw arbitrary arrows between two states, which will be just like **mutable state `class` with `var`s**.

Reviewer 3: Ah yes, I missed the state lifting from the reducer passed in the init, to the reducer passed to the state machine.

Reviewer 5: > If we allow `Update` to be passed in, it means we are going to draw arbitrary arrows between two states, which will be just like **mutable state `class` with `var`s**.

That's exactly my concern too.

Author: > mutable state class with vars

That's not entirely correct, because mutation will emit new State

Author: Well I guess it's a trade off of dropping some actions that do not give much value. To some extent we just bind a component to particular sub state if only thing it does it updating its value ¯\_(ツ)_/¯

Author: But if the emitting event does something else e.g triggering another feedback or does some validation of input than we still are going to do it via `Action.someChange(String)`.

Reviewer 3: @sergdort is it the goal to update the state via the KeyPath conveniences?

Reviewer 2: can we address @mluisbrown and @inamiy concern by leaving handling of mutation in each view model rather than doing that in the base class? essentially mutation is a pair of key path and value, so why not to bundle it in one generic even? this seems to be the main idea of the suggestion, isn't it?

Reviewer 2: It also will hurt debugging if handling of all mutations is done in one central place, wouldn't it?

Reviewer 4: My two cents for this solution is to add **`SubState`** instead, and **disallow `Mutation` that receives current state as input** (since current-state referencing i.e. `Reducer` is a complicated mechanism that should only be handled by state machine, not each caller's sending).

```swift
struct State {
    var subState: SubState // Mutation-aware
    var others: ...  // not Mutation-aware
    
    // Explicitly define sub-state for mutation.
    struct SubState {
    	...
    }
}

enum Update {
    case event(Event)
    
    // NOTE: Directly passing a new sub-state WITHOUT referring current sub-state.
    // Behaviour is mostly the same as SwiftUI's direct state binding.
    case syncSubState(SubState)
}
```

This literally means **`SubState` will also behave as sub-action**,
which then allows us to not draw arbitrary state transition arrows, so it's safer.

Reviewer 4: Ah pls forget above.
`SubState` won't be fully able to behave as sub-action, since struct and enum are not the same but dual.

Author: @ilyapuchka 
> It also will hurt debugging if handling of all mutations is done in one central place, wouldn't it?

You would need to add it in every view model

Reviewer 2: but all the mutations are handled here in the base class, no code to handle key path mutations would be in the view models, which seems to be the point of proposal. so considering debugging the only point where we will be able to set breakpoint will be this switch and it will hit it on all view models, there will be no way to debug a single view model conveniently.

>       feedbacks: feedbacks.map { $0.mapEvent(Update.event) }

Reviewer 2: what is this `mapEvent`?

Reviewer 5: See here: https://github.com/sergdort/CombineFeedback/blob/master/Sources/CombineFeedbackUI/ViewModel.swift#L62-L66

Author: I guess in terms of point free it's a form of pull back 🤔

Reviewer 3: A form of pull back would be on the state (versus the event):

```
func mapEvent<U>(_ f: (Event -> U) -> Feedback<State, U> {

}
```

versus

```
func contraMapEvent<U>(_ f: (State -> U) -> Feedback<U, Event> {

}
```

        .appending(self.input.feedback)
    )
  }

  open func send(action: Action) {
    self.input.observer(.event(actionTransformer(action)))
  }

  open func mutate<V>(keyPath: WritableKeyPath<State, V>, value: V) {
    self.input.observer(.mutation(Mutation(keyPath: keyPath, value: value)))
  }

  private enum Update {
    case event(Event)
    case mutation(Mutation)
  }

>   private struct Mutation {

Reviewer 2: this will probably benefit a little bit from @dynamicCallable and this new proposal https://github.com/apple/swift-evolution/blob/master/proposals/0253-callable.md This way instead of `mutation.mutate(state)` we can just have `mutation(state)`.

```

2. Have one entity that both contains the State and can dispatch the action e.g [StateSnapshot](https://github.com/andersio/Cycler/blob/master/View/BoundView.swift#L34)
```swift
//@dynamicMemberLookup
public struct StateSnapshot<T: FeedbackLoopType> {
  public let state: T.State
  private weak var feedbackLoop: T?

  public init(state: T.State, feedbackLoop: T) {
    self.state = state
    self.feedbackLoop = feedbackLoop
  }

  public subscript<U>(dynamicMember keyPath: KeyPath<T.State, U>) -> U {
    get {
      return state[keyPath: keyPath]
    }
  }

  public func send(action: T.Action) {
    feedbackLoop?.send(action: action)
  }

  public func mutate<V>(keyPath: WritableKeyPath<T.State, V>, value: V) {
    feedbackLoop?.mutate(keyPath: keyPath, value: value)
  }
}
```
Then our render would look like:

```swift
public protocol Renderer {
  associatedtype FeedbackLoop: FeedbackLoopType
  associatedtype Appearance: BoxAppearance
  associatedtype Config

  init(appearance: Appearance, config: Config)

  func render(snapshot: StateSnapshot<FeedbackLoop>) -> AnyRenderable
}
```

This will allow us easier bind VM to a component e.g

```swift
return Component(textDidChange: {

>    //This will still couse rerendering but we would not need to create an Event for it

Reviewer: ```suggestion
   // this will still cause rerendering but we would not need to create an Event for it
```

    snapshot.mutate(\.userName, value: $0)
})
```

## Impact on the existing codebase

> This is additive no impact can be bundled together with new `RenderableViewController`

Reviewer: Can we rephrase this? Sounds confusing imho

Author: We have a `RenderableViewController` which is suppose to replace `BoxViewController`
